File: hora/horoscope/dhasa/graha/kaala.py
""" Kaala Dhasa """
from hora.panchanga import drik
from hora import utils, const
_kaala_dhasa_life_span = 120 # years
def _dhasa_progression_and_periods(jd,place):
    previous_day_sunset_time = drik.sunset(jd-1, place)[0]
    today_sunset_time = drik.sunset(jd, place)[0]
    today_sunrise_time = drik.sunrise(jd, place)[0]
    tomorrow_sunrise_time = 24.0+drik.sunrise(jd+1, place)[0]
    _,_,_,birth_time = utils.jd_to_gregorian(jd)
    df = abs(today_sunset_time - today_sunrise_time)/6.0
    nf1 = abs(today_sunrise_time-previous_day_sunset_time)/6.0
    nf2 = abs(tomorrow_sunrise_time-today_sunset_time)/6.0
    #print('df',df,'nf1',nf1,'nf2',nf2)
    dawn_start = today_sunrise_time-nf1; dawn_end=today_sunrise_time+nf1
    #print('dawn',dawn_start,dawn_end)
    day_start = dawn_end; day_end = today_sunset_time-nf1
    #print('day',day_start,day_end)
    dusk_start = day_end ; dusk_end = today_sunset_time+nf2
    #print('dusk',dusk_start,dusk_end)
    yday_night_start = -(previous_day_sunset_time+nf1); yday_night_end = today_sunrise_time-nf1
    tonight_start = today_sunset_time+nf2; tonight_end = tomorrow_sunrise_time-nf2
    #print('Night-Yday',yday_night_start,yday_night_end,'Night-today',tonight_start,tonight_end)
    # Night is before dawn_start and after dusk_end
    if birth_time > dawn_start and birth_time < dawn_end: # dawn
        kaala_period = 'Dawn'
        kaala_frac = (birth_time-dawn_start)/(dawn_end-dawn_start)
    elif birth_time > dusk_start and birth_time < dusk_end: # dusk
        kaala_period = 'Dusk'
        kaala_frac = (birth_time-dusk_start)/(dusk_end-dusk_start)
    elif birth_time > day_start and birth_time < day_end: # Day
        kaala_period = 'Day'
        kaala_frac = (birth_time-day_start)/(day_end-day_start)
    elif birth_time > yday_night_start and birth_time < yday_night_end: # yday-night
        kaala_period = 'YDay-Night'
        kaala_frac = (birth_time-yday_night_start)/(yday_night_end-yday_night_start)
    elif birth_time > tonight_start and birth_time < tonight_end: # yday-night
        kaala_period = 'ToNight'
        kaala_frac = (birth_time-tonight_start)/(tonight_end-tonight_start)
    #print(birth_time,'kaala period',kaala_period,'kaala fraction',kaala_frac)
    _kaala_dhasa_life_span_first_cycle = _kaala_dhasa_life_span*kaala_frac
    _dhasas1 = [(p+1)*_kaala_dhasa_life_span_first_cycle/45.0 for p in range(9)]
    # Second Cycle
    _kaala_dhasa_life_span_second_cycle = _kaala_dhasa_life_span - _kaala_dhasa_life_span_first_cycle
    _dhasas2 = [(p+1)*_kaala_dhasa_life_span_second_cycle/45.0 for p in range(9)]
    
    return _dhasas1,_dhasas2
def get_dhasa_antardhasa(dob,tob,place,divisional_chart_factor=1,years=1,months=1,sixty_hours=1,include_antardhasa=False):
    jd_at_dob = utils.julian_day_number(dob, tob)
    jd_years = drik.next_solar_date(jd_at_dob, place, years=years, months=months,sixty_hours=months)
    dhasas_first, dhasas_second = _dhasa_progression_and_periods(jd_years, place)
    dhasa_info = []
    start_jd = jd_years
    for dhasa_lord in range(9):
        _dhasa_duration = dhasas_first[dhasa_lord]
        if include_antardhasa:
            for bhukthi_lord in range(9):
                _bhukthi_duration = (bhukthi_lord+1)*_dhasa_duration/45.0
                y,m,d,h = utils.jd_to_gregorian(start_jd)
                dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
                dhasa_info.append((dhasa_lord,bhukthi_lord,dhasa_start,round(_bhukthi_duration,2)))
                start_jd += _bhukthi_duration * const.sidereal_year
        else:
            y,m,d,h = utils.jd_to_gregorian(start_jd)
            dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
            dhasa_info.append((dhasa_lord,dhasa_start,round(_dhasa_duration,2)))
            start_jd += _dhasa_duration * const.sidereal_year
    # Second Cycle
    for dhasa_lord in range(9):
        _dhasa_duration = dhasas_second[dhasa_lord]
        if include_antardhasa:
            for bhukthi_lord in range(9):
                _bhukthi_duration = (bhukthi_lord+1)*_dhasa_duration/45.0
                y,m,d,h = utils.jd_to_gregorian(start_jd)
                dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
                dhasa_info.append((dhasa_lord,bhukthi_lord,dhasa_start,round(_bhukthi_duration,2)))
                start_jd += _bhukthi_duration * const.sidereal_year
        else:
            y,m,d,h = utils.jd_to_gregorian(start_jd)
            dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
            dhasa_info.append((dhasa_lord,dhasa_start,round(_dhasa_duration,2)))
            start_jd += _dhasa_duration * const.sidereal_year
    return dhasa_info
if __name__ == "__main__":
    from hora.tests import pvr_tests
    pvr_tests.kaala_test()

File: pyjhora-web/src/core/dhasa/graha/kaala.ts
/**
 * Kaala Dasha System
 * Ported from PyJHora kaala.py
 *
 * Time-based dasha system that divides the day into 4 kaala periods:
 * Dawn, Day, Dusk, Night
 * Total: 120 years split into two cycles based on birth kaala
 */

import {
  PLANET_NAMES_EN,
  SIDEREAL_YEAR,
  TROPICAL_YEAR
} from '../../constants';
import { sunrise, sunset } from '../../ephemeris/swe-adapter';
import type { Place } from '../../types';
import { julianDayToGregorian } from '../../utils/julian';

// ============================================================================
// TYPES
// ============================================================================

export interface KaalaDashaPeriod {
  lord: number;
  lordName: string;
  startJd: number;
  startDate: string;
  durationYears: number;
}

export interface KaalaBhuktiPeriod {
  dashaLord: number;
  bhuktiLord: number;
  bhuktiLordName: string;
  startJd: number;
  startDate: string;
  durationYears: number;
}

export interface KaalaResult {
  kaalaType: number; // 0=Dawn, 1=Day, 2=Dusk, 3=Night
  kaalaTypeName: string;
  kaalaFraction: number;
  mahadashas: KaalaDashaPeriod[];
  bhuktis?: KaalaBhuktiPeriod[];
}

// ============================================================================
// CONSTANTS
// ============================================================================

const KAALA_LIFE_SPAN = 120; // years
const YEAR_DURATION = SIDEREAL_YEAR;

const KAALA_TYPE_NAMES = ['Dawn', 'Day', 'Dusk', 'Night'];

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

function formatJdAsDate(jd: number): string {
  const { date, time } = julianDayToGregorian(jd);
  const pad = (n: number) => Math.abs(n).toString().padStart(2, '0');
  const hour12 = time.hour % 12 || 12;
  const ampm = time.hour < 12 ? 'AM' : 'PM';
  const yearStr = date.year < 0 ? `${Math.abs(date.year)} BC` : date.year.toString();
  return `${yearStr}-${pad(date.month)}-${pad(date.day)} ${pad(hour12)}:${pad(time.minute)}:${pad(time.second)} ${ampm}`;
}

/**
 * Approximate next_solar_date from Python drik.py.
 * Python uses inverse_lagrange + iterative solar longitude search for precision;
 * here we approximate by advancing JD by the tropical year fraction.
 * When years=1, months=1, sixtyHours=1 (defaults), returns jd unchanged (matching Python).
 * TODO: Implement full next_solar_date with solar longitude search once drik.ts has
 * solarLongitude + inverse_lagrange support.
 */
function nextSolarDateApprox(
  jd: number,
  _place: Place,
  years: number = 1,
  months: number = 1,
  sixtyHours: number = 1
): number {
  if (years === 1 && months === 1 && sixtyHours === 1) return jd;
  // Approximate: advance by the tropical year fraction (matches Python's jd_extra logic)
  const jdExtra = Math.floor(
    ((years - 1) + (months - 1) / 12 + (sixtyHours - 1) / 144) * TROPICAL_YEAR
  );
  return jd + jdExtra;
}

/**
 * Calculate kaala type and fraction based on birth time
 * Divides the day into 4 periods: Dawn, Day, Dusk, Night
 * Each period is further divided into 6 parts
 */
function calculateKaalaProgression(jd: number, place: Place): {
  kaalaType: number;
  kaalaFraction: number;
  firstCyclePeriods: number[];
  secondCyclePeriods: number[];
} {
  // Get sunrise/sunset times
  const previousDaySunset = sunset(jd - 1, place).localTime;
  const todaySunset = sunset(jd, place).localTime;
  const todaySunrise = sunrise(jd, place).localTime;
  const tomorrowSunrise = 24.0 + sunrise(jd + 1, place).localTime;

  // Calculate day and night fractions (1/6 of each period)
  const dayFraction = Math.abs(todaySunset - todaySunrise) / 6.0;
  const nightFraction1 = Math.abs(todaySunrise - previousDaySunset) / 6.0;
  const nightFraction2 = Math.abs(tomorrowSunrise - todaySunset) / 6.0;

  // Define period boundaries
  const dawnStart = todaySunrise - nightFraction1;
  const dawnEnd = todaySunrise + nightFraction1;
  const dayStart = dawnEnd;
  const dayEnd = todaySunset - nightFraction1;
  const duskStart = dayEnd;
  const duskEnd = todaySunset + nightFraction2;
  const yesterdayNightStart = -(previousDaySunset + nightFraction1);
  const yesterdayNightEnd = todaySunrise - nightFraction1;
  const tonightStart = todaySunset + nightFraction2;
  const tonightEnd = tomorrowSunrise - nightFraction2;

  // Get birth time from JD
  const { time } = julianDayToGregorian(jd);
  const birthTime = time.hour + time.minute / 60 + time.second / 3600;

  // Determine kaala type and fraction
  let kaalaType: number;
  let kaalaFraction: number;

  if (birthTime > dawnStart && birthTime < dawnEnd) {
    // Dawn
    kaalaType = 0;
    kaalaFraction = (birthTime - dawnStart) / (dawnEnd - dawnStart);
  } else if (birthTime > duskStart && birthTime < duskEnd) {
    // Dusk
    kaalaType = 2;
    kaalaFraction = (birthTime - duskStart) / (duskEnd - duskStart);
  } else if (birthTime > dayStart && birthTime < dayEnd) {
    // Day
    kaalaType = 1;
    kaalaFraction = (birthTime - dayStart) / (dayEnd - dayStart);
  } else if (birthTime > yesterdayNightStart && birthTime < yesterdayNightEnd) {
    // Yesterday's night (early morning before dawn)
    kaalaType = 3;
    kaalaFraction = (birthTime - yesterdayNightStart) / (yesterdayNightEnd - yesterdayNightStart);
  } else if (birthTime > tonightStart && birthTime < tonightEnd) {
    // Tonight
    kaalaType = 3;
    kaalaFraction = (birthTime - tonightStart) / (tonightEnd - tonightStart);
  } else {
    // Default to day if unable to determine
    kaalaType = 1;
    kaalaFraction = 0.5;
  }

  // Calculate dasha periods based on kaala fraction
  // First cycle: 9 periods, each duration = (lord_index + 1) * total_duration / 45
  const firstCycleLifeSpan = KAALA_LIFE_SPAN * kaalaFraction;
  const firstCyclePeriods = Array.from({ length: 9 }, (_, i) =>
    (i + 1) * firstCycleLifeSpan / 45.0
  );

  // Second cycle
  const secondCycleLifeSpan = KAALA_LIFE_SPAN - firstCycleLifeSpan;
  const secondCyclePeriods = Array.from({ length: 9 }, (_, i) =>
    (i + 1) * secondCycleLifeSpan / 45.0
  );

  return {
    kaalaType,
    kaalaFraction,
    firstCyclePeriods,
    secondCyclePeriods
  };
}

// ============================================================================
// MAIN FUNCTION
// ============================================================================

/**
 * Get Kaala Dasha data
 * @param jd - Julian Day Number (birth time)
 * @param place - Place data
 * @param options - Calculation options
 * @returns Kaala dasha result with mahadashas and optional bhuktis
 */
export function getKaalaDashaBhukti(
  jd: number,
  place: Place,
  options: {
    includeBhuktis?: boolean;
    years?: number;
    months?: number;
    sixtyHours?: number;
  } = {}
): KaalaResult {
  const { includeBhuktis = true, years = 1, months = 1, sixtyHours = 1 } = options;

  // Apply solar date adjustment (Python: drik.next_solar_date)
  const jdAdjusted = nextSolarDateApprox(jd, place, years, months, sixtyHours);

  const {
    kaalaType,
    kaalaFraction,
    firstCyclePeriods,
    secondCyclePeriods
  } = calculateKaalaProgression(jdAdjusted, place);

  const mahadashas: KaalaDashaPeriod[] = [];
  const bhuktis: KaalaBhuktiPeriod[] = [];

  let startJd = jdAdjusted;

  // Process both cycles
  const cycles = [
    { periods: firstCyclePeriods, fraction: kaalaFraction },
    { periods: secondCyclePeriods, fraction: 1 - kaalaFraction }
  ];

  for (const cycle of cycles) {
    for (let dashaLord = 0; dashaLord < 9; dashaLord++) {
      const durationYears = cycle.periods[dashaLord]!;

      if (includeBhuktis) {
        // Two sub-cycles for bhuktis within each dasha
        const bhuktiCycles = [cycle.fraction, 1 - cycle.fraction];

        for (const bhuktiCycleFraction of bhuktiCycles) {
          const cycleDuration = bhuktiCycleFraction * durationYears;

          for (let bhuktiLord = 0; bhuktiLord < 9; bhuktiLord++) {
            const bhuktiDuration = (bhuktiLord + 1) * cycleDuration / 45.0;
            const bhuktiLordName = PLANET_NAMES_EN[bhuktiLord] ?? `Planet ${bhuktiLord}`;

            bhuktis.push({
              dashaLord,
              bhuktiLord,
              bhuktiLordName,
              startJd,
              startDate: formatJdAsDate(startJd),
              durationYears: bhuktiDuration
            });

            startJd += bhuktiDuration * YEAR_DURATION;
          }
        }
      } else {
        const lordName = PLANET_NAMES_EN[dashaLord] ?? `Planet ${dashaLord}`;

        mahadashas.push({
          lord: dashaLord,
          lordName,
          startJd,
          startDate: formatJdAsDate(startJd),
          durationYears
        });

        startJd += durationYears * YEAR_DURATION;
      }
    }
  }

  // If bhuktis were generated, create mahadashas from them
  if (includeBhuktis && bhuktis.length > 0) {
    // Group bhuktis by dasha lord to create mahadashas
    let currentDashaLord = -1;
    let currentDashaStart = jdAdjusted;

    for (let i = 0; i < bhuktis.length; i++) {
      const bhukti = bhuktis[i]!;
      if (bhukti.dashaLord !== currentDashaLord) {
        if (currentDashaLord !== -1 && i > 0) {
          // Calculate duration for previous dasha
          const prevBhuktis = bhuktis.filter(b => b.dashaLord === currentDashaLord);
          const totalDuration = prevBhuktis.reduce((sum, b) => sum + b.durationYears, 0);

          mahadashas.push({
            lord: currentDashaLord,
            lordName: PLANET_NAMES_EN[currentDashaLord] ?? `Planet ${currentDashaLord}`,
            startJd: currentDashaStart,
            startDate: formatJdAsDate(currentDashaStart),
            durationYears: totalDuration
          });
        }
        currentDashaLord = bhukti.dashaLord;
        currentDashaStart = bhukti.startJd;
      }
    }

    // Add the last dasha
    if (currentDashaLord !== -1) {
      const lastBhuktis = bhuktis.filter(b => b.dashaLord === currentDashaLord);
      const totalDuration = lastBhuktis.reduce((sum, b) => sum + b.durationYears, 0);

      mahadashas.push({
        lord: currentDashaLord,
        lordName: PLANET_NAMES_EN[currentDashaLord] ?? `Planet ${currentDashaLord}`,
        startJd: currentDashaStart,
        startDate: formatJdAsDate(currentDashaStart),
        durationYears: totalDuration
      });
    }
  }

  return {
    kaalaType,
    kaalaTypeName: KAALA_TYPE_NAMES[kaalaType] ?? 'Unknown',
    kaalaFraction,
    mahadashas,
    bhuktis: includeBhuktis ? bhuktis : undefined
  };
}
